[SPARK-33573][SHUFFLE][YARN] Shuffle server side metrics for Push-based shuffle (#37638)
apache/spark: docs/monitoring.md

- shuffle-server.usedDirectMemory
- shuffle-server.usedHeapMemory

Below shuffle service server-side metrics are specific to the Push-Based Shuffle

> 

Author: Sure, below is the preview after the change, PTAL.
![Screen Shot 2023-01-10 at 12 15 47 PM](https://user-images.githubusercontent.com/31781684/211652893-14f01968-642c-4965-962b-af0c82d486db.png)

- shuffle-server.usedDirectMemory
- shuffle-server.usedHeapMemory

> - Notes: below shuffle service server-side metrics are specific to the Push-Based Shuffle
> (with `spark.shuffle.push.enabled` set as true on the client side, and with the server side
> flag `spark.shuffle.push.server.mergedShuffleFileManagerImpl` set as the appropriate
> org.apache.spark.network.shuffle.MergedShuffleFileManager implementation for the Push-Based
> Shuffle to be enabled)

Reviewer: These metrics are not related to client side configurations, as if push is not enabled in client side, these metrics will still be shown up, but all 0.  Once server side configuration is set correctly to MergedShuffleFileManager, these metrics will be shown up.

```suggestion
- **notes:** The metrics below are specific to push-based shuffle and only emitted when 
  `spark.shuffle.push.server.mergedShuffleFileManagerImpl` is configured as 
  "org.apache.spark.network.shuffle.RemoteBlockPushResolver".
```
Please refer to other notes format in this file. You do need to follow the format there.

Author: I removed the client side config in the description. As for the format for note, I'm following the format at L1408, L1399, L1392, and L1384. This is the preview after the change:
![Screen Shot 2023-01-10 at 4 23 19 PM](https://user-images.githubusercontent.com/31781684/211689574-adf92b25-b0b7-406d-9dc4-847e64a09afd.png)
Let me know if you have further input, thanks @zhouyejoe !

  - blockAppendCollisions - the number of shuffle push blocks collided in shuffle services
    as another block for the same reduce partition were being written
  - lateBlockPushes - the number of shuffle push blocks that are received in shuffle service
    after the specific shuffle merge has been finalized
  - blockBytesWritten - the size of the pushed block data written to file in bytes
  - deferredBlockBytes - the size of the current deferred block parts buffered in memory
  - deferredBlocks - the number of the current deferred block parts buffered in memory

>   - staleBlockPushes - the number of stale shuffle block push requests

Reviewer: Remove the indent. These metrics should be within the same level as other metrics in shuffle service.

Author: Fixed.

- shuffle-server.usedDirectMemory
- shuffle-server.usedHeapMemory

Note: applies to the shuffle service when the server side flag
`spark.shuffle.push.server.mergedShuffleFileManagerImpl` set as the appropriate
org.apache.spark.network.shuffle.MergedShuffleFileManager implementation for the Push-Based Shuffle

- blockAppendCollisions - the number of shuffle push blocks collided in shuffle services
  as another block for the same reduce partition were being written
- lateBlockPushes - the number of shuffle push blocks that are received in shuffle service
  after the specific shuffle merge has been finalized
- blockBytesWritten - the size of the pushed block data written to file in bytes
- deferredBlockBytes - the size of the current deferred block parts buffered in memory
- deferredBlocks - the number of the current deferred block parts buffered in memory
- staleBlockPushes - the number of stale shuffle block push requests
- ignoredBlockBytes - the size of the pushed block data that are ignored after the shuffle

>   file is finalized or when a request is for a duplicate block

Reviewer 2: nit: 
Can we reorder this so that we go from the common metrics to the more corner cases ?
For example, something like:

* blockBytesWritten
* blockAppendCollisions
* lateBlockPushes
* staleBlockPushes
* deferredBlocks
* deferredBlockBytes
* ignoredBlockBytes

(If this is not the right order, please feel free to reorder them)

Author: Thanks for the suggestion, I further put the `staleBlockPushes` to the middle of `deferredBlockBytes ` and `ignoredBlockBytes`, as a stale block push appears in the case of indeterminate stage retries, which I feel is less common to deferred block. Let me know if that's not right. Thanks.

Author: Also adjust a bit after @zhouyejoe's suggestion, preview of the corresponding section:
![Screen Shot 2023-01-11 at 2 50 26 PM](https://user-images.githubusercontent.com/31781684/211935150-cc71545b-41ad-4a0c-bc3c-901ef2dd8b5f.png)

Reviewer: LGTM.

Reviewer 2: Can we add a new line separation from the earlier list (before the `note: the metrics ...`) ?
All metrics following the note are related to push based shuffle - would be good to visually group them together.

Author: Sounds good. Current preview:
![Screen Shot 2023-01-11 at 6 31 34 PM](https://user-images.githubusercontent.com/31781684/211961919-efa4426a-df6f-43b7-8e07-32a455b2bb5f.png)
